File: hackerrank/problem solving/algorithms/2 - implementation/48 - halloween sale/halloween_sale.cs
// https://www.hackerrank.com/challenges/halloween-sale/problem?isFullScreen=true

int[] arr = ReadNumbers();
int price = arr[0];
int discount = arr[1];
int minPrice = arr[2];
int budget = arr[3];
Console.WriteLine(HowManyGamesCanBeBought(price, discount, minPrice, budget));

int[] ReadNumbers() => Console.ReadLine()!.Split(' ').Select(int.Parse).ToArray();

// T: O(budget / minPrice)
// S: O(1) extra space
int HowManyGamesCanBeBought(int price, int discount, int minPrice, int budget)
{
    int count = 0;
    while (budget >= price)
    {
        count++;
        budget -= price;
        price = Math.Max(price - discount, minPrice);
    }
    return count;
}

File: hackerrank/problem solving/algorithms/2 - implementation/48 - halloween sale/halloween_sale.py
# https://www.hackerrank.com/challenges/halloween-sale/problem?isFullScreen=true

def main():
    price, discount, min_price, budget = read_numbers()
    print(how_many_games_can_be_bought(price, discount, min_price, budget))


def read_numbers():
    return list(map(int, input().split()))


def how_many_games_can_be_bought(price, discount, min_price, budget):
    count = 0
    while budget >= price:
        count += 1
        budget -= price
        price = max(price - discount, min_price)
    return count


if __name__ == '__main__':
    main()
